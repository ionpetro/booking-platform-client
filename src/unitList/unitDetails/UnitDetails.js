import Message from '../../components/Message/Message.vue';
import UnitBook from './unitBook/UnitBook.vue';
import UnitService from '../unit.service';

const unitService = new UnitService();

export default {
  components: {
    UnitBook,
    'v-message': Message
  },
  props: ['selectedUnitId'],
  data() {
    return {
      unit: '',
      loading: false,
      message: ''
    };
  },
  mounted() {
    this.fetch();
  },
  methods: {
    async fetch() {
      this.loading = true;
      try {
        this.unit = await unitService.getUnitById(this.selectedUnitId);
        this.message = '';
      } catch (error) {
        this.unit = '';
        this.message = error.message;
      } finally {
        this.loading = false;
      }
    }
  }
};
