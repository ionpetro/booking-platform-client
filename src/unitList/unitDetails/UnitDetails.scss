.details {
  padding: 0 24px;
  height: 100vh;
  overflow-y: auto;

  &__error {
    position: absolute;
    z-index: 1;
    margin: 50px auto;
    left: 0;
    right: 0;
  }

  &__container {
    height: 100vh;
    display: flex;
    justify-content: center;
    align-items: center;

    &__loader {
      display: block;
    }
  }

  &__characteristics {
    display: flex;
    justify-content: flex-start;
    align-items: center;

    div {
      display: flex;
      span {
        padding: 4px 20px 4px 4px;
      }
    }
  }

  &__img__container {
    margin-left: -24px;
    width: 780px;
    height: 340px;

    @media #{$tablet} {
      max-width: 100vw;
      max-height: 100%;
    }
    @media #{$mobile} {
      height: 200px;
    }
  }

  h3 {
    margin-top: 24px;
  }

  h5 {
    margin-top: 32px;
    margin-bottom: 24px;
  }

  &__amenities {
    margin-top: -24px; // override margin from flex
    display: flex;
    flex-wrap: wrap;
    justify-content: flex-start;
    align-items: center;
  }
  &__amenity {
    display: flex;
    flex-direction: column;
    align-items: center;

    img {
      margin: 24px 8px 10px 0px;
    }

    @media #{$tablet} {
      width: 33%;
    }
    @media #{$mobile} {
      width: 50%;
      img {
        width: 156px;
        height: 112px;
      }
    }
    @media only screen and (max-width: 360px) {
      width: 100%;
    }
  }
  &__benefits {
    margin-top: -5px;
    display: flex;
    flex-wrap: wrap;
    justify-content: flex-start;

    div {
      display: flex;
      align-items: center;
      width: 33%;
      height: 50px;

      @media #{$tablet} {
        width: 50%;
      }
      @media #{$mobile} {
        width: 100%;
      }

      img {
        padding-right: 14px;
      }
    }
  }

  &__crater {
    img {
      max-width: 100%;
      max-height: 100%;
    }
  }

  &--bottomSpacing {
    // space on the bottom to fit the book container
    margin-top: 100px;
  }

  @media #{$mobile} {
    padding: 0 16px;
  }

  &__book {
    position: absolute;
    // drawer width
    width: 780px;
    // full block according to container paddings
    margin-left: -24px;
    bottom: 0;
    height: 80px;
    background: #ffffff;
    box-shadow: 0 -4px 15px rgba(159, 159, 159, 0.25);

    @media only screen and (max-width: 780px) {
      // adapt the block according to container paddings
      width: 100%;
    }

    @media #{$mobile} {
      margin-left: -16px;
    }
  }
}
