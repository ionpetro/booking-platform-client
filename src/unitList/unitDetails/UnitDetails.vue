<template>
  <div class="details">
    <transition name="fade">
      <v-message class="details__error" :type="'error'" v-if="message">{{ message }}</v-message>
    </transition>
    <div v-if="loading" class="details__container">
      <beat-loader
        class="details__container__loader"
        color="gray"
        :size="20"
        sizeUnit="px"
      ></beat-loader>
    </div>
    <div v-if="unit">
      <img class="details__img__container" :src="unit.image" :alt="`unit image ${unit.image}`" />
      <!-- Title -->
      <h3>{{ unit.crater }}, {{ unit.address }}</h3>
      <!-- Basic Characteristics -->
      <div class="details__characteristics">
        <div v-if="unit.capsules">
          <img src="@/assets/icons/Capsules.svg" alt="Capsules" /><span> {{ unit.capsules }}</span>
        </div>
        <div v-if="unit.sterilizers">
          <img src="@/assets/icons/Sterilizers.svg" alt="sterilizers" /><span>
            {{ unit.sterilizers }}</span
          >
        </div>
        <div v-if="unit.size">
          <img src="@/assets/icons/Size.svg" alt="Size" />
          <span> {{ unit.size }} mf</span>
        </div>
        <div v-if="unit.distance">
          <img src="@/assets/icons/Portals.svg" alt="Portals" />
          <span>{{ unit.distance }} portals</span>
        </div>
      </div>
      <!-- Amenities -->
      <h5>Amenities</h5>
      <div v-if="unit.amenities" class="details__amenities">
        <div
          class="details__amenity"
          v-for="(amenity, index) of unit.amenities"
          :key="`amenity%${index}`"
        >
          <img :src="require(`@/assets/images/${amenity}.png`)" :alt="amenity" />
          <span>{{ amenity }}</span>
        </div>
      </div>
      <div v-else>No amenities</div>
      <!-- Crater Description -->
      <h5>Crater</h5>
      <div v-if="unit.imageCrater" class="details__crater">
        <img :src="unit.imageCrater" alt="Crater Image" />
        <p>{{ unit.descriptionCrater }}</p>
      </div>
      <div v-else>No crater details</div>
      <!-- Benefits -->
      <h5>Crater benefits</h5>
      <div v-if="unit.benefits" class="details__benefits">
        <div
          class="details__benefit"
          v-for="(benefit, index) of unit.benefits"
          :key="`benefit%${index}`"
        >
          <img :src="require(`@/assets/icons/${benefit}.svg`)" :alt="benefit" />
          <span class="body--large">{{ benefit }}</span>
        </div>
      </div>
      <div v-else>No crater benefits</div>
      <div class="details--bottomSpacing"></div>
      <!-- Book Layout -->
      <div class="details__book">
        <UnitBook @onErrorMessage="bookingError" :unit="unit"></UnitBook>
      </div>
    </div>
  </div>
</template>

<script src="./UnitDetails.js"></script>

<style lang="scss" src="./UnitDetails.scss" scoped></style>
