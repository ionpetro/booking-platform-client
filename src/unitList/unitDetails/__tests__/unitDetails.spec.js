import flushPromises from 'flush-promises';
import {shallowMount} from '@vue/test-utils';
import unitMock from './mocks/unit.mock';
import UnitDetails from '../UnitDetails.vue';
import UnitService from '../../unit.service';

jest.mock('../../unit.service', () => jest.fn());
const mockResp = unitMock;
describe('UnitDetails', () => {
  let wrapper;

  beforeEach(() => {
    UnitService.mockClear();
  });

  afterEach(() => {
    wrapper.destroy();
  });

  describe('when loaded', () => {
    it('renders data correctly', async () => {
      UnitService.prototype.getUnitById = jest.fn(() => Promise.resolve(mockResp));
      wrapper = shallowMount(UnitDetails, {
        propsData: {
          selectedUnitId: 1
        }
      });

      await flushPromises();
      expect(wrapper.vm.$data.unit).toEqual(unitMock);
      expect(wrapper.element).toMatchSnapshot();
    });

    it('handles fetch data error', async () => {
      UnitService.prototype.getUnitById = jest.fn(() => Promise.reject(new Error('error')));
      wrapper = shallowMount(UnitDetails, {
        propsData: {
          selectedUnitId: 1
        }
      });

      await flushPromises();
      expect(wrapper.vm.$data.message).toEqual('error');
      expect(wrapper.element).toMatchSnapshot();
    });
  });
});
