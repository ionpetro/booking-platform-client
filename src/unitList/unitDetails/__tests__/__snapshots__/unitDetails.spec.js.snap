// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UnitDetails when loaded handles fetch data error 1`] = `
<div
  class="details"
>
  <transition-stub
    name="fade"
  >
    <v-message-stub
      class="list__error"
      type="error"
    >
      error
    </v-message-stub>
  </transition-stub>
   
  <!---->
   
  <!---->
</div>
`;

exports[`UnitDetails when loaded renders data correctly 1`] = `
<div
  class="details"
>
  <transition-stub
    name="fade"
  >
    <!---->
  </transition-stub>
   
  <!---->
   
  <div>
    <img
      alt="unit image image"
      class="details__img__container"
      src="image"
    />
     
    <h3>
      name, 1
    </h3>
     
    <div
      class="details__characteristics"
    >
      <div>
        <img
          alt="Capsules"
          src="@/assets/icons/Capsules.svg"
        />
        <span>
           1
        </span>
      </div>
       
      <!---->
       
      <div>
        <img
          alt="Size"
          src="@/assets/icons/Size.svg"
        />
         
        <span>
           1 mf
        </span>
      </div>
       
      <div>
        <img
          alt="Portals"
          src="@/assets/icons/Portals.svg"
        />
         
        <span>
          1 portals
        </span>
      </div>
    </div>
     
    <h5>
      Amenities
    </h5>
     
    <div
      class="details__amenities"
    >
      <div
        class="details__amenity"
      >
        <img
          alt="Cleaning robot"
          src=""
        />
         
        <span>
          Cleaning robot
        </span>
      </div>
      <div
        class="details__amenity"
      >
        <img
          alt="Extremely fast Wi-Fi"
          src=""
        />
         
        <span>
          Extremely fast Wi-Fi
        </span>
      </div>
      <div
        class="details__amenity"
      >
        <img
          alt="Sterilizing equipment"
          src=""
        />
         
        <span>
          Sterilizing equipment
        </span>
      </div>
      <div
        class="details__amenity"
      >
        <img
          alt="Personal assistant"
          src=""
        />
         
        <span>
          Personal assistant
        </span>
      </div>
      <div
        class="details__amenity"
      >
        <img
          alt="Automatic heat"
          src=""
        />
         
        <span>
          Automatic heat
        </span>
      </div>
      <div
        class="details__amenity"
      >
        <img
          alt="Comfortable capsule"
          src=""
        />
         
        <span>
          Comfortable capsule
        </span>
      </div>
      <div
        class="details__amenity"
      >
        <img
          alt="Oxygen system"
          src=""
        />
         
        <span>
          Oxygen system
        </span>
      </div>
      <div
        class="details__amenity"
      >
        <img
          alt="Cooking robot"
          src=""
        />
         
        <span>
          Cooking robot
        </span>
      </div>
    </div>
     
    <h5>
      Crater benefits
    </h5>
     
    <div
      class="details__benefits"
    >
      <div
        class="details__benefit"
      >
        <img
          alt="Automatic heating"
          src=""
        />
         
        <span
          class="body--large"
        >
          Automatic heating
        </span>
      </div>
      <div
        class="details__benefit"
      >
        <img
          alt="Fitness room"
          src=""
        />
         
        <span
          class="body--large"
        >
          Fitness room
        </span>
      </div>
      <div
        class="details__benefit"
      >
        <img
          alt="Private portal cabin"
          src=""
        />
         
        <span
          class="body--large"
        >
          Private portal cabin
        </span>
      </div>
      <div
        class="details__benefit"
      >
        <img
          alt="Internal pool"
          src=""
        />
         
        <span
          class="body--large"
        >
          Internal pool
        </span>
      </div>
      <div
        class="details__benefit"
      >
        <img
          alt="Pet friendly"
          src=""
        />
         
        <span
          class="body--large"
        >
          Pet friendly
        </span>
      </div>
    </div>
     
    <h5>
      Crater
    </h5>
     
    <div
      class="details__crater"
    >
      <img
        alt="Crater Image"
        src="image"
      />
       
      <p>
        desc
      </p>
    </div>
     
    <div
      class="details--bottomSpacing"
    />
     
    <div
      class="details__book"
    >
      <unitbook-stub
        unit="[object Object]"
      />
    </div>
  </div>
</div>
`;
